### PR TITLE
feat: use iptables filter traffics,only forward to allowip and port.

### PR DESCRIPTION
**Pull request by laoshancun**
Enhance Security by Restricting Traffic Forwarding to Specified AllowIPs and Ports

## Description

This PR significantly improves the security posture of the application by implementing the following changes:

1. Removed the default behavior that forwards all traffic through the VPN tunnel
2. Implemented traffic restriction logic to only forward traffic related to the specified AllowIPs
3. Added support for port-specific traffic restrictions within individual AllowIP entries
4. Refactored IP processing functions by extracting and reusing common utilities

The implementation includes:
- Moving the `isIPv6` and `parseIpAndPort` functions to a shared utility module
- Updating client traffic restriction plugin to use these shared utilities
- Modifying client configuration generation to properly handle IP addresses with port specifications
- Implementing iptables/ip6tables rules that restrict traffic to only the explicitly allowed IPs and ports
image

image

## Motivation and Context

The primary motivation for this change is to enhance security by implementing the principle of least privilege. Previously, the default configuration forwarded all traffic through the VPN, which could potentially expose internal networks or sensitive services.

Key security benefits include:

1. **Granular Access Control**: Users can now precisely specify which IPs and ports should be accessible through the VPN
2. **Reduced Attack Surface**: Limiting traffic to only necessary destinations decreases potential attack vectors
3. **Better Compliance**: The ability to restrict traffic aligns with common security compliance requirements
4. **Improved Network Segmentation**: Administrators can more effectively segment network access based on client needs

This change addresses a fundamental security concern while maintaining flexibility through the AllowIPs configuration, which now supports port-level granularity.

## How has this been tested?

The security enhancements have been thoroughly tested through multiple verification methods:

1. **Functional Testing**: Verified that traffic is only forwarded to explicitly allowed IPs and ports
2. **Security Validation**: Confirmed that unauthorized traffic destinations are properly blocked
3. **Regression Testing**: Ensured that existing VPN functionality remains intact for authorized connections
4. **Edge Case Testing**: Tested various IP formats (IPv4, IPv6, CIDR notation) with and without port specifications

All tests were conducted in both development and production-like environments, confirming the security improvements work as expected without introducing new issues.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring (non-breaking change that improves code quality and maintainability)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/server/plugins/clientTrafficRestriction.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+declare module 'nitropack/types' {
+  interface NitroRuntimeHooks {
+    'wireguard:config': () => void;
+    'wireguard:start': () => void;
+  }
+}
+
+// 客户端流量限制插件
+export default defineNitroPlugin(async (nitroApp) => {
```
Comment (kaaax0815) [inline src/server/plugins/clientTrafficRestriction.ts]: i dont think that nitro is the right place for that

File: src/server/plugins/clientTrafficRestriction.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  console.log('Client Traffic Restriction Plugin loaded');
+
+  // 生成客户端特定的iptables规则
```
Comment (kaaax0815) [inline src/server/plugins/clientTrafficRestriction.ts]: remove non english comments

File: src/server/utils/WireGuard.ts [hunk 3]
```diff
@@ -1,18 +1,29 @@
...
+  }
+  // 注册事件监听器，监听客户端配置保存后的事件
   /**
    * Save and sync config
    */
   async saveConfig() {
     const wgInterface = await Database.interfaces.get();
     await this.#saveWireguardConfig(wgInterface);
     await this.#syncWireguardConfig(wgInterface);
+    await this.nitroApp.hooks.callHook('wireguard:config');
```
Comment (kaaax0815) [inline src/server/utils/WireGuard.ts]: this should be opt in

File: src/server/utils/wgHelper.ts [hunk 4]
```diff
@@ -101,6 +102,9 @@ PostDown = ${iptablesTemplate(hooks.postDown, wgInterface)}`;
       dnsServers.length > 0 ? `DNS = ${dnsServers.join(', ')}` : null;
 
     const extraLines = [dnsLine, ...hookLines].filter((v) => v !== null);
+    const allowIps = (client.allowedIps ?? userConfig.defaultAllowedIps).map(
```
Comment (kaaax0815) [inline src/server/utils/wgHelper.ts:155]: i dont quite see the reason for this. why would there be a be a port in allowedips? adding port specific allowed ips makes no sense to me

File: src/server/utils/ip.ts [hunk 5]
```diff
@@ -170,3 +170,44 @@ async function getIpInformation() {
 export const cachedGetIpInformation = cacheFunction(getIpInformation, {
   expiry: 15 * 60 * 1000,
 });
+
+// 检查IP地址是否为IPv6
+export const isIPv6 = (ip: string): boolean => {
```
Comment (kaaax0815) [inline src/server/utils/ip.ts]: there is already the `is-ip` package installed

**Review by kaaax0815 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Awesome, restricting access to specific ports is very useful for me.

 I hope it can be approved through a PR as soon as possible

**Comment by kubakubakuba** [comment]
Is there any progress on this? Seems like a genuinely usefull feature...

**Comment by kaaax0815** [comment]
the author of the pr seems to be inactive. someone could take over of course

**Comment by laoshancun** [comment]
Anyone really need this feature?

**Comment by ghost** [comment]
Yes, this feature is very necessary for me. 
Previously, by default, all port data for IPv4 or IPv6 would be forwarded, which is insecure. It is very necessary to limit the data outgoing from specified ports. 

This is a good feature

File: src/server/utils/ip.ts (hunk 5 above)
Comment (laoshancun) [reply src/server/utils/ip.ts]: ok

File: src/server/utils/wgHelper.ts (hunk 4 above)
Comment (laoshancun) [reply src/server/utils/wgHelper.ts:155]: As discussed below, the AllowedIPs parameter incorporates the port filtering feature, enabling more fine-grained access permission control.

File: src/server/plugins/clientTrafficRestriction.ts (hunk 1 above)
Comment (laoshancun) [reply src/server/plugins/clientTrafficRestriction.ts]: Any idea?

File: src/server/utils/wgHelper.ts (hunk 4 above)
Comment (kaaax0815) [reply src/server/utils/wgHelper.ts:155]: i don't think this should be added

File: src/server/plugins/clientTrafficRestriction.ts (hunk 1 above)
Comment (kaaax0815) [reply src/server/plugins/clientTrafficRestriction.ts]: e.g. in https://github.com/wg-easy/wg-easy/blob/eb6b96c0f18405a7fd948e7add77fc34b392a9d4/src/server/utils/WireGuard.ts#L61

Comment (laoshancun) [reply src/server/plugins/clientTrafficRestriction.ts]: That's fine.

File: src/server/utils/wgHelper.ts (hunk 4 above)
Comment (laoshancun) [reply src/server/utils/wgHelper.ts:155]: Cause wireguard only support ip or CIDR,we should remove port from AllowedIPs.

**Comment by lanrat** [comment]
I've just submitted a PR for a very similar feature in #2418.

**Comment by laoshancun** [comment]
Closed,see also #2418 .

**Comment by kaaax0815** [comment]
Closing in favor of #2418
